Add offline PDC API

Lulu13022002 opened the pull request:
Closes https://github.com/PaperMC/Paper/issues/7288
<details>
<summary>Old description</summary>
Plugin can now gets the offline persistent data of a player. ~The current behavior is like the Machine's Advancement PR:
All setters re-save the data directly into the disk~
Note also that if spigot config disablePlayerDataSaving is enabled, the data will not be written
</details>

Plugin can now gets the offline persistent data of a player with a read-only view.

Lulu13022002 committed:
Add offline PDC API

lynxplay commented on patches/server/0925-Add-Offline-PDC-API.patch:
```diff
@@ -0,0 +1,99 @@
++        if(player != null) return player.getPersistentDataContainer();
++
++        CompoundTag result = getData();
++        if(result == null) return null; // never called
```
Am I missing something ?
This if definitely is possible if a server requests an offline player e.g. based on uuid that has not joined before.

```java
final OfflinePlayer dinnerbone = Bukkit.getOfflinePlayer(UUID.fromString("61699b2e-d327-4a01-9f1e-0ea8c3f06bc6"));
final PersistentDataContainer persistentDataContainer = dinnerbone.getPersistentDataContainer();
```
would yield a null pdc instance (unless dinnerbone played on your server before) which contradicts the `@NotNull` annotation on https://jd.papermc.io/paper/1.19/org/bukkit/persistence/PersistentDataHolder.html#getPersistentDataContainer().

Lulu13022002 replied:
Well not really i thought that the Bukkit.getOfflinePlayer call returns null for unknown player but i was wrong

lynxplay commented on patches/server/0925-Add-Offline-PDC-API.patch:
```diff
@@ -0,0 +1,99 @@
++        CompoundTag result = getData();
++        if(result == null) return null; // never called
++
++        org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer pdc = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(org.bukkit.craftbukkit.entity.CraftEntity.DATA_TYPE_REGISTRY) {
```
Might also want to override the `clear` method. Looking at #7505, that method might be used soon.
Could also just be part of the other PR tho.

lynxplay commented on patches/server/0925-Add-Offline-PDC-API.patch:
```diff
@@ -0,0 +1,99 @@
+ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+     private static PermissibleBase perm;
+-    private static final CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new CraftPersistentDataTypeRegistry();
++    public static final CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new CraftPersistentDataTypeRegistry(); // Paper private -> public
```
If you want, this could be in the access transformers instead of a manual change.
Up to you, I don't think we really have much of a preference.

Machine-Maker replied:
I'd leave it as a change here. I think we are trying to move away from making everything an AT, leaving that for larger things, or things that need to change visibility across inheritance.

lynxplay commented on patches/api/0387-Add-Offline-PDC-API.patch:
```diff
@@ -0,0 +1,38 @@
++     *
++     * @return the persistent metadata container
++     */
++    @NotNull
```
The javadocs should properly mention the side effects of an offline players persistent data container.
E.g. that every write operation (remove, set) triggers disk IO.

Lulu13022002 replied:
Also do you think a closable persistent data can be better to avoid rewrite for each operation

lynxplay commented on patches/server/0925-Add-Offline-PDC-API.patch:
```diff
@@ -0,0 +1,99 @@
+ 
+     }
++    // Paper start
++    public void save(java.util.UUID uuid, CompoundTag tag) {
```
Unsure how to properly do this, this method is pretty much just a copy of the existing `PlayerDataStorage#save(Player)`.
Could reduce duplicate code here.

Machine-Maker commented on patches/server/0933-Add-Offline-PDC-API.patch:
```diff
@@ -0,0 +1,212 @@
++        }
++        return false;
++    }
++    // Paper end
```
Be more strict with your `// Paper start` and `// Paper end` comments. Most of the time, they should only be surrounding actual diff, not a big chunk with several line changes inside it.

Lulu13022002 replied:
Resolved, removed some useless diff

lynxplay commented on patches/api/0391-Add-Offline-PDC-API.patch:
```diff
@@ -0,0 +1,65 @@
++     * @return the persistent metadata container
++     */
++    @NotNull
++    org.bukkit.persistence.PersistentDataContainer getPersistentDataContainer();
```
Any reason this is not just returning `io.papermc.paper.persistence.OfflinePersistentDataContainer` then ? If you are supposed to always cast it ?

Lulu13022002 replied:
Like i have said before this interface is also used for the online player (CraftPlayer) and so an offline pdc doesn't make sense for online player

lynxplay replied:
Right, the permanent issue of players extending offline players, I forgot.
This is some ugly API ngl, the fact that Player would then return you a PDC that cannot be cast but offline players do is pretty fucked.

lynxplay commented on patches/api/0391-Add-Offline-PDC-API.patch:
```diff
@@ -0,0 +1,65 @@
++/**
++ * A {@link AutoCloseable} Persistent Data Container
++ */
++public interface OfflinePersistentDataContainer extends PersistentDataContainer, AutoCloseable {
```
Are we expecting this interface to be used again ?
I am unsure when, but the general concept of a PDC that does not "write" until closed might come up again, at which point this naming may be too specific.

Lulu13022002 replied:
Yes ideally all data that are retrieved by a single file like the advancements and the statistics for the offline player must use a similar things. For the pdc i agree with you but i have no idea of a good name.

lynxplay commented on patches/server/0933-Add-Offline-PDC-API.patch:
```diff
@@ -0,0 +1,215 @@
++        return false; // Paper
+     }
++    // Paper start
++    public boolean searchAvailableData(CompoundTag c) {
```
This method name basically says nothing. what available data is it searching ? This method is also more of a contains/has logic, not really a search.

Lulu13022002 replied:
Yes, you've right what about hasBukkitValues/hasBukkitPersistentData name ?

lynxplay reviewed (commented):
Just some thoughts. In general, this PR got kinda more ugly than initially set out to be because it aims to support PDCs for players that have never joined and hence we need all this mess.

lynxplay committed:
Rebase

lynxplay committed:
Move getOrDefault to impl

Moves the getOrDefault impl back to the impl type to correctly "mirror"
the copy process from upstreams PDC interface.

lynxplay reviewed (approved):
Generally, LGTM.
Moved the impl of getOrDefault back to the impl, as that is where it lives in upstreams impl.

Only thing left to discuss pre-merge imo is a potential javadocs warning about how the impl works, given the pdc getter itself does not do IO but every query to it does, which might throw people off compared to the usual behaviour of OfflinePlayer, which does all its IO on the method itself and then returns a snapshot (e.g. location or metadata lists)

lynxplay committed:
Add caution notice to offline pdc view

lynxplay committed:
Jetbrains annotations

lynxplay committed:
Import me this lulu

Lulu13022002 committed:
last fix
